juspay/hyperswitch
# feat(payment_charges): add support for collecting and refunding charges on payments

## Type of Change



## Description

Detailed description - https://github.com/juspay/hyperswitch/issues/4659

This PR adds functionality to create charges on payment intents using Stripe
- create `charges` on payment_intent
- store `charge_id` in payment_attempt
- use `charge_id` for refunds
   - Direct
   ```
    "charges": {
        "charge_id": "ch_3PJDz8Ihl7EEkW0O16Wetnl5",
        "revert_platform_fee": true
    }
    ```
   - Destination
   ```
    "charges": {
        "charge_id": "ch_3PJDz8Ihl7EEkW0O16Wetnl5",
        "revert_platform_fee": true,
        "revert_transfer": true
    }
_Note_ - charges on mandates is not added in this PR.

### Additional Changes





## Motivation and Context



## How did you test it?

Tested locally using postman collection.

**Steps for testing**
- Make sure Stripe Connect is enabled on your Stripe account
- Onboard a Standard user by initiating an account link from Stripe's dashboard
- Make sure merchant account is created and Stripe is added as a payment connector
- Create a direct charge on payment by adding below field in `/payment_intents` request
```
"charges": {
        "charge_type": "direct",
        "fees": 123,
        "transfer_account_id": "acct_1PDftAIhl7EEkW0O"
}
```
- Create a destination charge on payment by adding below field in `/payment_intents` request
```
"charges": {
        "charge_type": "destination",
        "fees": 123,
        "transfer_account_id": "acct_1PDftAIhl7EEkW0O"
}
```

### Direct charges
- ref - https://docs.stripe.com/connect/direct-charges
- these help in directly creating charges when customer is paying to the connected account
- charges are created, money is split (Stripe fees + reseller’s fees) and the remaining amount is transferred to connected account’s balance

### Destination charges
- ref - https://docs.stripe.com/connect/destination-charges
- charges are created on the reseller’s account
- reseller decides whether some of all of those funds are to be transferred to the connected account
- reseller’s account is used for debiting the cost of Stripe fees, refunds and chargebacks
- only supported when both reseller and connected accounts belong to the same country



## Checklist
